strapi/strapi
# [Blocks] Fix content not updating when switching locales

### Why is it needed?

There were 2 bugs related to i18n and blocks:

- switching locales did not update the content in the admin
- "fill in from locale" did not work

Both were due to the fact that the Slate component is uncontrolled (see https://github.com/ianstormtaylor/slate/issues/4612). So it was not aware that the value was being updated somewhere else by Redux.

### What does it do?

Since we can't use Slate as a controlled component, my workaround is to count the total updates to `value`, vs the `value` updates that were triggered by Slate. When they're different, it means the `value` was updated from Redux (e.g. when switching locales). In this case it forces a rerender of the Slate component, which will now use the latest `value` version.

I extracted the logic to a custom hook because I didn't want this hack to be in the way of the rest of the component. It also made it easier to document it (comments) since everything related is in 1 place.

### How to test it?

Use a content type that has localization enabled (and at least 2 locales) and a blocks attribute.

Save content in 1 locale, switch to the other. Use "fill in from locale". Switch back and forth between locales. The content you see in the editor should always be in sync with the active locale, without needing to refresh.
